oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java

```diff
@@ -105,20 +110,20 @@ public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2Authentic
 			throw new OAuth2AuthenticationException(oauth2Error, oauth2Error.toString());
 		}
 		RequestEntity<?> request = this.requestEntityConverter.convert(userRequest);
-		ResponseEntity<Map<String, Object>> response = getResponse(userRequest, request);
-		Map<String, Object> userAttributes = response.getBody();
+		ResponseEntity<OAuth2UserAuthority> response = getResponse(userRequest, request);
+		OAuth2UserAuthority userAuthority = response.getBody();
 		Set<GrantedAuthority> authorities = new LinkedHashSet<>();
-		authorities.add(new OAuth2UserAuthority(userAttributes));
+		authorities.add(userAuthority);
 		OAuth2AccessToken token = userRequest.getAccessToken();
 		for (String authority : token.getScopes()) {
 			authorities.add(new SimpleGrantedAuthority("SCOPE_" + authority));
 		}
-		return new DefaultOAuth2User(authorities, userAttributes, userNameAttributeName);
+		return new DefaultOAuth2User(authorities, userAuthority.getAttributes(), userNameAttributeName);
 	}
 
-	private ResponseEntity<Map<String, Object>> getResponse(OAuth2UserRequest userRequest, RequestEntity<?> request) {
+	private ResponseEntity<OAuth2UserAuthority> getResponse(OAuth2UserRequest userRequest, RequestEntity<?> request) {
```

Comment:
Why does this return `OAuth2UserAuthority` instead of `Map<String, Object>`? The UserInfo response returns a set of claim/attribute name/value pairs.

Reply:
If I remember correctly, this was done to align the created `OAuth2UserAuthorityHttpMessageConverter` with e.g. `OAuth2AccessTokenResponseHttpMessageConverter` in the same package. Additionally, I wasn't sure how to satisfy `supports(Class<?> clazz)` - so I ended up with this construct. I don't mind changing this to old value, as it's just a implementation detail. 

Before my change, the following was done with the returned `Map<String, Object> userAttributes`:

`authorities.add(new OAuth2UserAuthority(userAttributes));` - so it felt like a good abstraction, if I move it to the private method. 

Any preference?

Reply:
See [5.3.2. Successful UserInfo Response](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse):

> The UserInfo Claims MUST be returned as the members of a JSON object ...

`OAuth2UserAuthority` would not be the equivalent representation. We should use `OidcUserInfo`.

We can create a new `HttpMessageConverter`:

`public class OidcUserInfoHttpMessageConverter extends AbstractHttpMessageConverter<OidcUserInfo>`

Then we can configure it as follows:

```java
RestTemplate restTemplate = new RestTemplate(
		Collections.singletonList(
				new OidcUserInfoHttpMessageConverter()));
restTemplate.setErrorHandler(new OAuth2ErrorResponseErrorHandler());

DefaultOAuth2UserService userService = new DefaultOAuth2UserService();
userService.setRestOperations(restTemplate);

OidcUserService oidcUserService = new OidcUserService();
oidcUserService.setOauth2UserService(userService);
```

Makes sense?

Reply:
Thanks for your feedback! I haven't compared the spec with the used class - thanks for the input! I try to change it in the upcoming days.

Reply:
@jgrandja Adopted your suggestions and removed my new constructor for `DefaultOAuth2UserService`, because it can be easily done with your code shown above to customize it now.